tests/scattering1d/test_timefrequency_scattering.py

```diff
@@ -151,9 +151,111 @@ def test_frequency_scattering():
 
     # spinned=True
     X['coef'] = X['coef'].astype('complex64')
-    X['n'] = (-1, 4) # a mockup (n1, n2) pair from scattering1d_widthfirst
+    X['n'] = (-1, 4) # a mockup (n1, n2) pair from time_scattering_widthfirst
     freq_gen = frequency_scattering(X, S.backend, jtfs.filters_fr,
         jtfs.oversampling_fr, jtfs.average_fr=='local', spinned=True)
     for Y_fr in freq_gen:
         assert Y_fr['coef'].shape[-1] == X['coef'].shape[-1]
         assert Y_fr['n'] == (4, Y_fr['n_fr'])
+
+
+def test_joint_timefrequency_scattering():
+    # Define scattering object
+    J = 8
+    J_fr = 3
+    shape = (4096,)
+    Q = 8
+    S = TimeFrequencyScatteringBase(
+        J=J, J_fr=J_fr, shape=shape, Q=Q, T=0, F=0, backend='numpy')
+    S.build()
+    S.create_filters()
+    assert not S.average
+    assert not S.average_fr
+    backend = kymatio.Scattering1D(
+        J=J, Q=Q, shape=shape, T=0, backend='numpy').backend
+
+    x = torch.zeros(shape)
+    x[shape[0]//2] = 1
+    x_shape = backend.shape(x)
+    batch_shape, signal_shape = x_shape[:-1], x_shape[-1:]
+    x = backend.reshape_input(x, signal_shape)
+    U_0_in = backend.pad(x, pad_left=S.pad_left, pad_right=S.pad_right)
+
+    filters = [S.phi_f, S.psi1_f, S.psi2_f]
+    jtfs_gen = joint_timefrequency_scattering(U_0_in, backend,
+        filters, S.oversampling, S.average=='local',
+        S.filters_fr, S.oversampling_fr, S.average_fr=='local')
+    path_keys = ['coef', 'j', 'n', 'n1_max', 'j_fr', 'n_fr', 'spin']
+
+    # Zeroth order
+    U_0 = next(jtfs_gen)
+    assert np.allclose(U_0['coef'], U_0_in)
+    assert U_0['n'] == ()
+    assert U_0['j'] == ()
+
+    # First and second order
+    S_jtfs = list(jtfs_gen)
+
+    # Check that all path keys are unique
+    ns = [path['n'] for path in S_jtfs]
+    assert len(ns) == len(set(ns))
+
+    # Check that order 1 comes before order 2
+    orders = [len(path['n']) for path in S_jtfs]
+    assert set(orders) == {1, 2}
+    assert orders == sorted(orders)
+
+    # Test first order
+    S1_jtfs = filter(lambda path: len(path['n'])==1, S_jtfs)
+    for path in S1_jtfs:
+        # Check path format
+        assert path.keys() == set(path_keys)
+
+        # Check that first-order spins are nonnegative
+        assert path['spin'] >= 0
+
+        # Check frequency padding
+        assert path['n1_max'] < S._N_padded_fr
+
+        # Check that first-order coefficients have the same temporal stride
+        stride = 2**max(S.log2_T - S.oversampling, 0)
+        assert path['coef'].shape[-1] == (S._N_padded // stride)
+
+        # Check that frequential stride works as intended
+        stride_fr = 2**max(path['j_fr'] - S.oversampling_fr, 0)
+        assert (path['coef'].shape[-2]*stride_fr) == S._N_padded_fr
```

Comment:
i noticed that this line is written differently from line 222, i don't see the need but i'm ok with it. the rest looks good to me!

Reply:
good point @lylyhan. wil normalize

Reply:
be49c37